### PR TITLE
Handle more scenarios for loop cloning

### PR DESCRIPTION
**Pull request by kunalspathak**
- Increasing loops that are incremented by more than 1 like "i += 2"

[TypeHashingAlgorithm.cs](https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/Common/TypeSystem/Common/TypeHashingAlgorithms.cs)

https://github.com/dotnet/runtime/blob/7bac4e8fa611eb02ac1bd27b97a8570cbf321437/src/coreclr/tools/Common/TypeSystem/Common/TypeHashingAlgorithms.cs#L44-L49

- Decreasing loops like "i--", "i -= 2", etc.

https://github.com/dotnet/runtime/blob/7bac4e8fa611eb02ac1bd27b97a8570cbf321437/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs#L50-L63

PriorityQueue's [Heapify](https://github.com/dotnet/runtime/blob/7bac4e8fa611eb02ac1bd27b97a8570cbf321437/src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs#L576) method.

https://github.com/dotnet/runtime/blob/7bac4e8fa611eb02ac1bd27b97a8570cbf321437/src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs#L588-L591

Contributes to https://github.com/dotnet/runtime/issues/65342

### CODE REVIEW

**Comment by kunalspathak** [comment]
Failures are related to https://github.com/dotnet/runtime/issues/67878

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

**Comment by BruceForstall** [comment]
You need to worry about increment/decrement overflowing a 32-bit integer and wrapping. E.g.:
```
public static int test_up_big(int[] a, int s)
    {
        int r = 0;
        int i;
        for (i = 1; i < s; i += 2147483647)
        {
            r += a[i];
        }
        return r;
    }
```
with your code will clone, but generate an access violation, not System.IndexOutOfRangeException.

In the case of `for ( ; i < x; i += c)`, `c` needs to be strictly less than `int.MaxValue - (Array.MaxLength - 1) + 1`, which is `0x7fffffff - 0x7fffffc7 + 2` = 0x3a = 58. The idea being that if `i` was the maximum allowed legal array index of 0x7fffffc6 and you added 0x3a, the result is negative (-2,147,483,648).

Similar applies to down counted loops.

**Comment by kunalspathak** [comment]
> public static int test_up_big(int[] a, int s)
>     {
>         int r = 0;
>         int i;
>         for (i = 1; i < s; i += 2147483647)
>         {
>             r += a[i];
>         }
>         return r;
>     }

Thanks for pointing that out. Let me try to fix that.

**Comment by kunalspathak** [comment]
@BruceForstall - I tried measuring some of the benchmarks that showed up in asmdiffs and they all show improvements.



| Faster                       | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| ---------------------------- | ---------:| ----------------:| ----------------:| --------:|
| BilinearTest.Interpol_Scalar |      1.23 |         12591.21 |         10245.69 |         |
| BilinearTest.Interpol_Vector |      1.08 |         27857.16 |         25908.72 |         |
| BenchmarksGame.FannkuchRedux_5.RunBench(n: 10, expectedSum: 38) |      1.06 |      16428450.00 |      15492476.47 |         |
| BenchmarksGame.ReverseComplement_1.RunBench |      1.13 |       1021397.94 |        903558.01 |         |
| ByteMark.BenchEmFloatClass |      1.05 |     771413700.00 |     734018150.00 |         |
| ByteMark.BenchEmFloat      |      1.03 |    3737632200.00 |    3641513150.00 |         |
| System.Collections.Tests.Perf_PriorityQueue<String, String>.HeapSort(Size: 100)  |     10.80 |        898710.07 |         83248.11 |         |
| System.Collections.Tests.Perf_PriorityQueue<String, String>.HeapSort(Size: 10)   |     10.44 |         39025.89 |          3739.26 |         |
| System.Collections.Tests.Perf_PriorityQueue<String, String>.HeapSort(Size: 1000) |     10.34 |      14949560.00 |       1445691.48 |         |
| System.Collections.Tests.Perf_PriorityQueue<Int32, Int32>.HeapSort(Size: 10)     |      1.04 |           203.82 |           195.96 |         |
| Benchstone.BenchI.HeapSort.Test                                                  |      1.03 |        294538.92 |        287004.51 |         |


| Slower                                                                      | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| --------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Collections.Tests.Perf_PriorityQueue<Guid, Guid>.HeapSort(Size: 100) |      1.02 |          9744.88 |          9900.55 |         |
| System.Collections.Tests.Perf_PriorityQueue<Guid, Guid>.HeapSort(Size: 10)  |      1.01 |           396.27 |           401.45 |         |

Here is an example asmdiff from `Heapify`:

image

**Comment by kunalspathak** [comment]
ping.

File: src/coreclr/jit/loopcloning.cpp [hunk 1]
```diff
@@ -1022,18 +1022,29 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
...
+        int stride = abs(loop->lpIterConst());
+
+        if (stride > 58)
         {
-            JITDUMP("> Stride %d is invalid\n", loop->lpIterConst());
+            // Array.MaxLength can have maximum of 0X7FFFFFC7 elements, so make sure
+            // the stride increament doesn't overflow or underflow the index. Hence,
+            // the maximum stride limit is set to
+            // (int.MaxValue - (Array.MaxLength - 1) + 1), which is
+            // (0X7FFFFFC7 - 0x7fffffc7 + 2) = 0x3a or 58.
```
Comment (BruceForstall) [inline src/coreclr/jit/loopcloning.cpp:1043]:
```suggestion
            // (0X7fffffff - 0x7fffffc7 + 2) = 0x3a or 58.
```

File: src/coreclr/jit/loopcloning.cpp [hunk 2]
```diff
@@ -1022,18 +1022,29 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
...
+        // We already know that this is either increasing or decreasing loop and the
+        // stride is (> 0) or (< 0). Here, just take the abs() value and check if it
+        // is beyond the limit.
+        int stride = abs(loop->lpIterConst());
+
+        if (stride > 58)
         {
-            JITDUMP("> Stride %d is invalid\n", loop->lpIterConst());
+            // Array.MaxLength can have maximum of 0X7FFFFFC7 elements, so make sure
+            // the stride increament doesn't overflow or underflow the index. Hence,
```
Comment (BruceForstall) [inline src/coreclr/jit/loopcloning.cpp:1040]: typo
```suggestion
            // the stride increment doesn't overflow or underflow the index. Hence,
```

File: src/coreclr/jit/loopcloning.cpp [hunk 3]
```diff
@@ -1022,18 +1022,29 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
...
+    if (GenTree::StaticOperIs(loop->lpTestOper(), GT_LT, GT_LE, GT_GT, GT_GE))
     {
-        // Stride conditions
-        if (loop->lpIterConst() <= 0)
+        // We already know that this is either increasing or decreasing loop and the
+        // stride is (> 0) or (< 0). Here, just take the abs() value and check if it
+        // is beyond the limit.
+        int stride = abs(loop->lpIterConst());
+
+        if (stride > 58)
```
Comment (BruceForstall) [inline src/coreclr/jit/loopcloning.cpp:1037]:
```suggestion
        if (stride >= 58)
```

File: src/coreclr/jit/compiler.hpp [hunk 4]
```diff
@@ -3498,6 +3498,27 @@ inline genTreeOps Compiler::LoopDsc::lpTestOper() const
 
 //-----------------------------------------------------------------------------
 
+inline bool Compiler::LoopDsc::lpIsIncreasingLoop() const
+{
+    // Increasing loop is the one that has "+=" increament operation and "< or <=" limit check.
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.hpp:3503]: typo
```suggestion
    // Increasing loop is the one that has "+=" increment operation and "< or <=" limit check.
```

File: src/coreclr/jit/loopcloning.cpp [hunk 5]
```diff
@@ -1022,18 +1022,29 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
     JITDUMP("------------------------------------------------------------\n");
     JITDUMP("Deriving cloning conditions for " FMT_LP "\n", loopNum);
 
-    LoopDsc*                         loop     = &optLoopTable[loopNum];
-    JitExpandArrayStack<LcOptInfo*>* optInfos = context->GetLoopOptInfo(loopNum);
+    LoopDsc*                         loop             = &optLoopTable[loopNum];
+    JitExpandArrayStack<LcOptInfo*>* optInfos         = context->GetLoopOptInfo(loopNum);
+    bool                             isIncreasingLoop = GenTree::StaticOperIs(loop->lpTestOper(), GT_LT, GT_LE);
```
Comment (BruceForstall) [inline src/coreclr/jit/loopcloning.cpp:1027]: Why is `isIncreasingLoop` not defined using `loop->lpIsIncreasingLoop()`?

File: src/coreclr/jit/loopcloning.cpp [hunk 6]
```diff
@@ -1107,15 +1148,22 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
...
 
-        // GT_LT loop test: limit <= arrLen
-        // GT_LE loop test: limit < arrLen
+        // Increasing loops
+        // GT_LT loop test: (start < end) ==> (end <= start)
+        // GT_LE loop test: (start <= end) ==> (end < start)
+        //
+        // Decreasing loops
+        // GT_GT loop test: (end > start) ==> (end <= start)
+        // GT_GE loop test: (end >= start) ==> (end < start)
```
Comment (BruceForstall) [inline src/coreclr/jit/loopcloning.cpp:1157]: These conditions don't make sense.

Shouldn't it be, e.g.:
```suggestion
        // GT_LT loop test: (start < end) ==> (end <= arrLen)
```

File: src/coreclr/jit/loopcloning.cpp [hunk 7]
```diff
@@ -1618,8 +1659,13 @@ bool Compiler::optIsLoopClonable(unsigned loopInd)
...
     }
 
-    if (!((GenTree::StaticOperIs(loop.lpTestOper(), GT_LT, GT_LE) && (loop.lpIterOper() == GT_ADD)) ||
-          (GenTree::StaticOperIs(loop.lpTestOper(), GT_GT, GT_GE) && (loop.lpIterOper() == GT_SUB))))
+    // TODO-CQ: Handle other loops like:
+    // - The ones whose limit operator is "==" or "!="
+    // - The incrementing operator is multiple and divide
+    // - The ones that are inverted are not handled here for cases like "i *= 2" because
+    //   they are converted to "i + i".
+    // bool isOtherLoop = (loop.lpIterOper() == GT_DIV) || (loop.lpIterOper() == GT_MUL);
```
Comment (BruceForstall) [inline src/coreclr/jit/loopcloning.cpp:1667]: Delete commented code?

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/loopcloning.cpp (hunk 6 above)
Comment (kunalspathak) [reply src/coreclr/jit/loopcloning.cpp:1157]: yes, I too was confused initially but forgot to revisit this.

File: src/coreclr/jit/loopcloning.cpp (hunk 5 above)
Comment (kunalspathak) [reply src/coreclr/jit/loopcloning.cpp:1027]: yeah, I extracted `lpIsIncreasingLoop()` recently. Will fix it.

**Review by BruceForstall — APPROVED** [review]
LGTM.

I think you should run a jitstress job before merging.

**Comment by kunalspathak** [comment]
Improvements on windows/x64 : https://github.com/dotnet/perf-autofiling-issues/issues/5072

**Pull request merged**